### ISSUE TITLE
Access functionality for MatrixFree on faces

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#8460]
Following the post https://groups.google.com/d/msgid/dealii/13c24a74-caad-4c04-9b9c-d68e7b874d4c%40googlegroups.com?utm_medium=email&utm_source=footer we should introduce some additional access functionality to `MatrixFree` similar to `MatrixFree::get_cell_iterator`:

1)   functions similar to get_cell_iterator for getting iterator to interior/exterior cell +  face index:
```c++
std::pair<DoFHandler< dim >::cell_iterator,  unsigned int> 
get_face_iterator(const unsigned int face_batch_number,
 	  	  const unsigned int vector_number, const bool interior=true,
	  	  const unsigned int fe_component=0)
```
2) getter for `MatrixFree::cell_level_index` so that the data from `get_face_info` could be used easily interpreted without manually filling `std::map`

**Comment by peterrum** [comment]
I have once implemented in a separate branch:
```cpp
std::array<unsigned int, VectorizedArrayType::n_array_elements>
FEEvaluationBase::
get_cell_ids() const;
```
which I am using in
```cpp
inline VectorizedArrayType
FEEvaluationBase::
read_cell_data(const AlignedVector<VectorizedArrayType> &array) const;
```
to get the ids for inner and exterior faces. I would recommend that I integrate this functionality into `MatrixFree` and use the info to setup the face iterator. @kronbichler that might be a quick fix to 1.

**Comment by mwichro** [comment]
This should be added to matrix_free.h: 

  /**
   * Return the cell iterator in deal.II speak to a interior/exterior cell of given face  in the
   * renumbering of this structure.  The the second element of the pair is face number so that 
   * the face iterator can be accessed:
   * pair.first()->face(pair.second() );
   * 
   *
   * Note that the face iterators in deal.II go through cells differently to
   * what the face/boundary loop of this class does. This is because several faces are
   * worked on together (vectorization), and since faces with neighbor cells on
   * different MPI processors need to be accessed at a certain time when
   * accessing remote data and overlapping communication with computation.
   */
  std::pair<typename DoFHandler<dim>::cell_iterator,  unsigned int> 
  get_face_iterator(const unsigned int face_batch_number,
   	  	  const unsigned int vector_number,
		  const bool interior=true,
  	  	  const unsigned int fe_component=0) const;

**Comment by mwichro** [comment]
And this the code that should be put into matrix_free.templates.h. 
I did not tested that, there may be some mistakes. ( I will  have to update my dealii first )

std::pair<typename DoFHandler<dim>::cell_iterator,  unsigned int>
MatrixFree<dim, Number>::get_face_iterator(const unsigned int face_batch_number,
 	  	  const unsigned int vector_number, const bool interior=true,
	  	  const unsigned int fe_component=0)const {

	std::pair<typename DoFHandler<dim>::cell_iterator,  unsigned int> result;

	  const unsigned int vectorization_length =
	    VectorizedArray<Number>::n_array_elements;
	  AssertIndexRange(fe_component, dof_handlers.n_dof_handlers);
//	  AssertIndexRange(face_batch_number, ???? );
	  AssertIndexRange(vector_number, n_active_entries_per_face_batch(face_batch_number));

	  const DoFHandler<dim> *dofh = nullptr;
	  if (dof_handlers.active_dof_handler == DoFHandlers::usual)
	    {
	      AssertDimension(dof_handlers.dof_handler.size(),
	                      dof_handlers.n_dof_handlers);
	      dofh = dof_handlers.dof_handler[fe_component];
	    }
	  else
	    {
	      Assert(false,
	             ExcMessage("Cannot return DoFHandler<dim>::cell_iterator "
	                        "for underlying DoFHandler!"));
	    }

	  const internal::MatrixFreeFunctions::FaceToCellTopology
	  		  < VectorizedArray< Number >::n_array_elements > face2cell_info=get_face_info(face);

	  const unsigned int cell_index= interior ?
			  face2cell_info.cells_interior[vector_number] :
			  face2cell_info.cells_exterior[element];

	  std::pair<unsigned int, unsigned int> index =
	    cell_level_index[cell_index];

	  result.first=typename DoFHandler<dim>::cell_iterator(&dofh->get_triangulation(),
	                                                 index.first,
	                                                 index.second,
	                                                 dofh);
	  result.second=interior ? face2cell_info.interior_face_no: face2cell_info.exterior_face_no ;
	  
	  return result;

}

**Comment by peterrum** [comment]
@kronbichler  Isn't this already closed by #8909?

**Comment by kronbichler** [comment]
Closed by #8909.
